[RESTWS-673] Scheduler REST Implementation

suthagar23 opened the pull request:
Created TaskDefinitionResource1_8 for the task REST implementation

dkayiwa commented:
Can you add some tests?

suthagar23 commented:
@dkayiwa Yah Sure @dkayiwa

gayanW commented:
I just ran the tests :) Tests in [TaskActionController1_8Test.java](https://github.com/openmrs/openmrs-module-webservices.rest/pull/293/files#diff-adf6af6ec197ebb5b06ac9203c55aded) fails.

suthagar23 commented:
@dkayiwa and @gayanW  I completed the TaskActionController1_8Test part. Now It's working without any failure :-)

dkayiwa commented on omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_8/TaskDefinitionResource1_8.java:
```diff
@@ -0,0 +1,184 @@
+	//	@Override
+	//	public DelegatingResourceDescription getUpdatableProperties() {
+	//		DelegatingResourceDescription description = new DelegatingResourceDescription();
+	//		description.addRequiredProperty("name");
```
Did you take a look at this? https://wiki.openmrs.org/display/docs/Java+Conventions#JavaConventions-Donotcommentoutcode

suthagar23 replied:
Removed those Comments

dkayiwa commented on omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_8/TaskDefinitionResource1_8.java:
```diff
@@ -0,0 +1,184 @@
+	 */
+	@Override
+	public TaskDefinition save(TaskDefinition taskDefinition) {
+		//Context.getSchedulerService().saveTaskDefinition(taskDefinition);
```
What is the above line for? https://wiki.openmrs.org/display/docs/Java+Conventions#JavaConventions-Donotcommentoutcode

suthagar23 replied:
Removed that line

dkayiwa commented on omod-common/src/main/java/org/openmrs/module/webservices/helper/TaskFactoryWrapper.java:
```diff
@@ -0,0 +1,89 @@
+				schedulerService.scheduleTask(task);
+			}
+			catch (SchedulerException e) {
+				throw new UnsupportedOperationException();
```
Does this really mean that the operation is not supported? The same applies elsewhere.

suthagar23 replied:
Removed unwanted Catch Expections

dkayiwa commented on omod-common/src/main/java/org/openmrs/module/webservices/helper/TaskFactoryWrapper.java:
```diff
@@ -0,0 +1,89 @@
+ */
+package org.openmrs.module.webservices.helper;
+
+import org.openmrs.scheduler.*;
```
Did you take a look at this? https://wiki.openmrs.org/display/docs/Java+Conventions#JavaConventions-Donotusewildcardimports

suthagar23 replied:
Refactored :-)

dkayiwa commented on omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_8/TaskActionResource1_8.java:
```diff
@@ -0,0 +1,224 @@
+			// ensure they're not trying to modify the REST module
+			for (TaskDefinition taskDef : taskDefinitions) {
+				// if they specified a module that's not loaded, it will show up here as null
+				if (taskDef == null) {
```
How does taskDefinitions become a collection of nulls?

dkayiwa commented on omod-1.8/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs1_8/TaskActionController1_8Test.java:
```diff
@@ -0,0 +1,145 @@
+	@Autowired
+	RestService restService;
+	
+	private TaskDefinition TestTask = new TaskDefinition(1, "TestTask", "TestTask Description",
```
Our convention is to use variable names that start with lower case.

suthagar23 replied:
Refactored

dkayiwa commented on omod-1.8/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs1_8/TaskActionController1_8Test.java:
```diff
@@ -0,0 +1,145 @@
+	
+	MockTaskFactoryWrapper mockTaskFactory = new MockTaskFactoryWrapper();
+	
+	String TestTaskMockJson = "{\"id\" : 1,\"name\":\"TestTask\",\"description\":\"TestTask Description\","
```
same as above for variable names.

suthagar23 replied:
Refactored

dkayiwa commented on omod-1.8/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs1_8/TaskActionController1_8Test.java:
```diff
@@ -0,0 +1,145 @@
+	
+	@Test
+	public void shouldStopTestTask() throws Exception {
+		System.out.println("shouldStopTestTask");
```
Take a look at this: https://wiki.openmrs.org/display/docs/Java+Conventions#JavaConventions-Donotprintout

suthagar23 replied:
Removed :-)

suthagar23 commented:
@dkayiwa  I have refactored the code about your comments

dkayiwa commented:
How about other comments that have nothing to do with refactoring?

dkayiwa commented on omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_8/TaskActionResource1_8.java:
```diff
@@ -0,0 +1,221 @@
+		
+		ServletContext servletContext = getServletContext(context);
+		
+		if (action.getAction().equals("SCHEDULETASK") || action.getAction().equals("RESCHEDULETASK")
```
Don't you already have constants for SCHEDULETASK, RESCHEDULETASK, etc?

suthagar23 replied:
Changed with constants from TaskAction

dkayiwa commented on omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_8/TaskActionResource1_8.java:
```diff
@@ -0,0 +1,221 @@
+				taskFactoryWrapper.reScheduleTask(taskDef);
+			}
+			catch (SchedulerException e) {
+				throw new UnsupportedOperationException();
```
I see so many UnsupportedOperationException. Does this mean if anything does wrong during these actions, it means that the rest call is not supported?

suthagar23 replied:
I couldn't throw the SchedulerException , because method implemented from Creatable Interface doesn't  allow to throw the SchedulerException. Else I wanted to change the Creatable Interface. So I decided to throw UnsupportedOperationException

suthagar23 replied:
Shall I throw ResponseException instead of UnsupportedOperationException?

dkayiwa replied:
Since you are not doing anything with the thrown exception, just do not catch it.

suthagar23 replied:
If I do not catch it, then want to throw it by methods. 
If I throw it from here, then want to throw it from Create method.
But I can't throw it because It is a overridden method. So I caught and throw like this. 
Are there any perfect methods for this?

suthagar23 replied:
@dkayiwa Shall I throw like this, 
throw new ResourceDoesNotSupportOperationException("SchedulerException");

suthagar23 replied:
@dkayiwa  If I throw the exception from here, then I can catch this from the angular REST request to indicate the user as well

Is it good to use like,
**throw ObjectNotFoundException("SchedulerException : Ca not execute the action")**

dkayiwa replied:
Do something along these lines:
throw new APIException("Errors occurred while rescheduling task", e);

suthagar23 replied:
Changed! Thanks @dkayiwa  :-)

dkayiwa commented on omod-1.8/src/test/java/org/openmrs/module/webservices/rest/web/MockTaskFactoryWrapper.java:
```diff
@@ -0,0 +1,65 @@
+		}
+		if (!contains) {
+			scheduledTasks.add(task.getName());
+		}
```
Did you intentionally decide to silently ignore a call for scheduling a task which does not exist?

suthagar23 replied:
Here, I am doing same as ModuleAction Test. I using mock collections to maintain registered and scheduled tasks. When REST call arrives at Resource, that will call these MockTaskFactoryWrapper class methods. Here I am mocking that tasks and assigning to the collections to check the status from Test Class.

dkayiwa commented on omod-1.8/src/test/java/org/openmrs/module/webservices/rest/web/MockTaskFactoryWrapper.java:
```diff
@@ -0,0 +1,65 @@
+		}
+		if (!contains) {
+			scheduledTasks.add(task.getName());
+		}
```
The same applies for others like task shutdown, reschedule, etc

dkayiwa commented on omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_8/TaskActionResource1_8.java:
```diff
@@ -0,0 +1,222 @@
+				taskFactoryWrapper.unloadTask(taskDef);
+			}
+			catch (SchedulerException e) {
+				throw new APIException("Errors occurred while shutdowning all tasks", e);
```
Is this exception message correct?

suthagar23 replied:
Ow sorry, Missed :-( 
Changed to "Errors occurred while deleting task".

dkayiwa commented:
@suthagar23 have you looked at the travis failure?

suthagar23 commented:
@dkayiwa That was mistake in the OpenMRS license header :-( 
Now I have fixed!

dkayiwa commented on omod-common/src/main/java/org/openmrs/module/webservices/helper/TaskFactoryWrapper.java:
```diff
@@ -0,0 +1,83 @@
+
+public class TaskFactoryWrapper {
+
+    private SchedulerService schedulerService = Context.getSchedulerService();
```
We do not recommend setting services on declaration because they may not be setup yet by the time the class is loaded.

suthagar23 replied:
So Shall I  create a service initialize function here and will call from TaskActionResource - Setup function to initiate the service?

dkayiwa replied:
setup function sounds cool.

suthagar23 replied:
So shall I initialize inside the constructor? 
```
    public TaskFactoryWrapper(){
        schedulerService = Context.getSchedulerService();
    }
```

dkayiwa replied:
Better than on declaration. As long as this class is not a spring bean.

suthagar23 replied:
I removed the declaration, and use  Context.getSchedulerService() in every place

dkayiwa commented on omod-common/src/main/java/org/openmrs/module/webservices/helper/TaskFactoryWrapper.java:
```diff
@@ -0,0 +1,83 @@
+
+import java.util.Collection;
+
+public class TaskFactoryWrapper {
```
What is the main reason for setting up this class?

suthagar23 replied:
I separated the Service actions from Resource and If I create the class like this,
Then only I can extend this class to MockTaskFactoryWrapper to mock the event for the tests.
(Actually, I followed the ModuleFactoryWrapper)

dkayiwa replied:
Don't you think the name should change? Aren't factories for creating instances of objects?

suthagar23 replied:
Class name has been changed to  TaskServiceWrapper

suthagar23 commented:
@dkayiwa I have changed the code about your comments
1. Changed the class name to TaskServiceWrapper
2. Removed service declaration, and use Context.getSchedulerService() in every place

dkayiwa commented on omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_8/TaskActionResource1_8.java:
```diff
@@ -0,0 +1,220 @@
+		setConvertedProperties(action, post, getCreatableProperties(), true);
+		
+		Collection<TaskDefinition> taskDefinitions;
+		if (action.isAllTasks() != null && action.isAllTasks()) {
```
Why do you test for action.isAllTasks() != null?

suthagar23 replied:
here for some actions, I will include `{ "allTasks"  : true }` in the REST body. So here I am checking that parameter for the confirmation. (Common actions only need this parameter, others can skip this)

dkayiwa replied:
Should isAllTasks() return null? I though it is a boolean function returning either true or false.

suthagar23 replied:
If the user's doesn't mention in the REST request, then It may be null. 
Like for this request, 

`{ "action" : "shutdowntask", "tasks" : ["taskA"] }`

dkayiwa replied:
By the way, why would the client have to mention allTasks? Why would i need to pass a list of tasks if am not interested in all of them?

dkayiwa commented on omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_8/TaskActionResource1_8.java:
```diff
@@ -0,0 +1,220 @@
+	}
+	
+	// Stop and start a scheduled task.
+	private void reScheduleTask(Collection<TaskDefinition> taskDefs) {
```
Since this is a collection of taskDefs. Can you add an s and change to reScheduleTasks?
The same applies to the other methods.

suthagar23 replied:
Sure :-)

dkayiwa commented on omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_8/TaskActionResource1_8.java:
```diff
@@ -0,0 +1,220 @@
+		}
+		ServletContext servletContext = getServletContext(context);
+		
+		if (action.getAction() == TaskAction.Action.SCHEDULETASK || action.getAction() == TaskAction.Action.RESCHEDULETASK
```
Here i would assign the result of getAction to a variable.

suthagar23 replied:
Changed

dkayiwa commented on omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_8/TaskActionResource1_8.java:
```diff
@@ -0,0 +1,220 @@
+				        + " on empty set of task definitions.");
+			}
+		}
+		if (action.isAllTasks() == null || !action.isAllTasks()) {
```
Like i said above, i do not find it API friendly for a boolean function to require checking for null. It should either be true or false.

suthagar23 replied:
I will default assign isAllTasks = false

dkayiwa replied:
Change its type too.

dkayiwa commented on omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_8/TaskActionResource1_8.java:
```diff
@@ -0,0 +1,220 @@
+		}
+		if (action.isAllTasks() == null || !action.isAllTasks()) {
+			for (TaskDefinition taskDef : taskDefinitions) {
+				if (taskDef == null) {
```
In what cases will taskDefinitions have a taskDef that is null?

suthagar23 replied:
if users sent an empty value of REST request, then It may happen? Am I correct? 

`{ "action" : "shutdowntask", "tasks" : ["taskA", ""] }`

dkayiwa replied:
Can you prove it?

suthagar23 replied:
I just put for the safety :-) . I will try and remove that if it's no need here. 
I think, mostly no need here !!!

dkayiwa replied:
Can it even ever happen? A collection with null?

dkayiwa commented on omod-common/src/main/java/org/openmrs/module/webservices/helper/TaskAction.java:
```diff
@@ -0,0 +1,60 @@
+	
+	private Action action;
+	
+	private Boolean allTasks;
```
Why do you make this nullable? Isn't more friendly to change the type to boolean and have it default to false?

suthagar23 replied:
okay I will do it like that way :-)

suthagar23 replied:
anyway I want to check again `isAllTasks == null` to change the boolean value from false to true?

dkayiwa replied:
How can it be null when the type is changed to boolean?

suthagar23 replied:
There will be no null type here anymore.
Default value or the Boolean will be False So I will check for that here.

dkayiwa commented on omod-common/src/main/java/org/openmrs/module/webservices/helper/TaskServiceWrapper.java:
```diff
@@ -0,0 +1,81 @@
+public class TaskServiceWrapper {
+	
+	public TaskDefinition getTaskById(String id) {
+		return Context.getSchedulerService().getTask(Integer.parseInt(id));
```
Shouldn't this be the task uuid?

suthagar23 replied:
here, It hasn't that much of function. I used taskName and TaskId instead of that

dkayiwa replied:
Do you know anywhere in our REST calls where clients pass integer id instead of uuid?

suthagar23 replied:
Nope,

but the problem is, It hasn't the function to getTask by UUID here. So Shall I want to implement that also?

dkayiwa replied:
That would mean modifying the core platform. In that case, just live it to get by name and integer id.

suthagar23 replied:
Okay, So I will just go through the name and id by now

dkayiwa commented on omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_8/TaskActionResource1_8.java:
```diff
@@ -0,0 +1,220 @@
+				break;
+			case RESCHEDULETASK:
+				reScheduleTask(taskDefinitions);
+				break;
```
I understand the actions to schedule, reschedule, delete, etc. But i do not understand the actions like onstartup and onshutdown. Does the user make a rest call and say on startup? For what purpose?

suthagar23 replied:
Those are default scheduler methods, 
OnStartup : Start all tasks that are scheduled to run on startup.
onShutDown : Stop all tasks and clean up the scheduler instance.

dkayiwa replied:
Can you describe an existing use case where a client will make a REST call as onStartup or onShutDown?

dkayiwa replied:
Putting it in a different way, let us do only the functionality that we actually need for our GSOC user interface work. No more. 😊

suthagar23 replied:
If he shut down some the tasks in the system and wanted to schedule the tasks which were active in the system startup, then he can just use onStartUp function.

If he want to shut down all tasks in the system, then use onShutdow in one click

dkayiwa replied:
Is that functionality needed for our GSOC user interface for managing tasks?

suthagar23 replied:
Actually, It's not needed for the GSoC User interface if I remove the Shutdown all tasks button from there.
I will remove it by now and complete after the GSoC time :-)

dkayiwa commented on omod-common/src/main/java/org/openmrs/module/webservices/helper/TaskServiceWrapper.java:
```diff
@@ -0,0 +1,81 @@
+	}
+	
+	public boolean isTaskStarted(TaskDefinition task) {
+		return task.getStarted();
```
I do not see value in a method like this. Why doesn't the caller just do task.getStarted()?

suthagar23 replied:
I wished to use TaskServiceWrapper for all kind of works. So Implement all the functionalities through TaskServiceWrapper. 
Is it bad, then I will remove :-)

dkayiwa replied:
Generally, more methods is more code to maintain. So let us not write any extra code unless there is real value that it adds. Just wrapping a method call in another method call does not seem to add any value. Does it? 😊

suthagar23 replied:
Yah absolutely. It will be just faking the Code :blush: 
I will do required changes and remove unwanted methods here

dkayiwa commented on omod-common/src/main/java/org/openmrs/module/webservices/helper/TaskServiceWrapper.java:
```diff
@@ -0,0 +1,81 @@
+	}
+	
+	public boolean isTaskStopped(TaskDefinition task) {
+		return !isTaskStarted(task);
```
Same here.

dkayiwa commented on omod-1.8/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_8/TaskDefinitionResource1_8Test.java:
```diff
@@ -0,0 +1,79 @@
+import org.slf4j.LoggerFactory;
+
+/**
+ * Created by sutha on 8/16/2017.
```
For the above, did you see this? https://wiki.openmrs.org/display/docs/Java+Conventions#JavaConventions-Attribution

suthagar23 replied:
Sorry for that, Removed :-(

dkayiwa commented on omod-common/src/main/java/org/openmrs/module/webservices/helper/TaskServiceWrapper.java:
```diff
@@ -0,0 +1,81 @@
+	}
+	
+	public void scheduleTask(TaskDefinition task) throws SchedulerException {
+		if (!Context.getSchedulerService().getTaskByName(task.getName()).getStarted()) {
```
After taking a second look, i think we do not need this call at all. All those check of ensure that a service is not started before schedule, etc, belong to the service layer API.
Your role is to simply call Context.getSchedulerService().scheduleTask(task) and that is all.

suthagar23 replied:
Okay :-). I agree with you
No need of much conditions at this point

dkayiwa commented on omod-common/src/main/java/org/openmrs/module/webservices/helper/TaskServiceWrapper.java:
```diff
@@ -0,0 +1,81 @@
+	
+	public void shutDownTask(TaskDefinition task) throws SchedulerException {
+		if (Context.getSchedulerService().getTaskByName(task.getName()).getStarted()) {
+			Context.getSchedulerService().shutdownTask(task);
```
Same here as above. So let us get rid of this whole class.

suthagar23 replied:
Agreed, I have removed all the conditional statements here and just called Context.getSchedulerService().shutdownTask(task);

dkayiwa commented on omod-1.8/src/test/java/org/openmrs/module/webservices/rest/web/MockTaskServiceWrapper.java:
```diff
@@ -0,0 +1,66 @@
+	@Override
+	public void scheduleTask(TaskDefinition task) throws SchedulerException {
+		boolean contains = false;
+		for (String taskDef : scheduledTasks) {
```
Why don't we just get rid of this class too??
Shouldn't this just call? Context.getSchedulerService().scheduleTask(task)

suthagar23 replied:
This is the MockTaskServiceWrapper class, I am using this for testing purpose. 
So at that point, I haven't any registered tasks to the schedule by Context.getSchedulerService().scheduleTask(task). 

I followed the MockModuleWrapper Idea here to do the tests.

dkayiwa replied:
Oh i see!
If it is only used for testing purposes and hence in src/test, then you are correct to have it stay. 👍

dkayiwa commented on omod-1.8/src/test/java/org/openmrs/module/webservices/rest/web/MockTaskServiceWrapper.java:
```diff
@@ -0,0 +1,66 @@
+	@Override
+	public void shutDownTask(TaskDefinition task) throws SchedulerException {
+		for (String taskDef : scheduledTasks) {
+			if (taskDef.equals(task.getName())) {
```
Shouldn't this also just call Context.getSchedulerService(). shutDownTask(task)
Same for the others.

suthagar23 replied:
It is also using this for testing purpose :blush: . So I just mocked it

dkayiwa commented on omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_8/TaskDefinitionResource1_8.java:
```diff
@@ -0,0 +1,167 @@
+	/**
+	 * @see org.openmrs.module.webservices.rest.web.resource.impl.DelegatingCrudResource#doSearch(org.openmrs.module.webservices.rest.web.RequestContext)
+	 *      A query string and/or a tag uuid can be passed in; if both are passed in, returns an
+	 *      intersection of the results; excludes retired locations
```
retired locations?

dkayiwa commented on omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_8/TaskDefinitionResource1_8.java:
```diff
@@ -0,0 +1,167 @@
+	
+	/**
+	 * @see org.openmrs.module.webservices.rest.web.resource.impl.DelegatingCrudResource#doSearch(org.openmrs.module.webservices.rest.web.RequestContext)
+	 *      A query string and/or a tag uuid can be passed in; if both are passed in, returns an
```
Is this a copied and pasted description from somewhere else?

dkayiwa commented on omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_8/TaskActionResource1_8.java:
```diff
@@ -0,0 +1,220 @@
+	}
+	
+	private ServletContext getServletContext(RequestContext context) {
+		return context.getRequest().getSession().getServletContext();
```
Can you remind me what this servlet context is used for?

suthagar23 replied:
Removed, It's not required here :-)

dkayiwa commented on omod-1.8/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs1_8/TaskActionController1_8Test.java:
```diff
@@ -0,0 +1,137 @@
+	}
+	
+	@Test
+	public void shouldStartTestTask() throws Exception {
```
shouldStart or shouldSchedule?

suthagar23 replied:
renamed to shouldScheduleTestTask();

dkayiwa commented on omod-1.8/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs1_8/TaskActionController1_8Test.java:
```diff
@@ -0,0 +1,137 @@
+	}
+	
+	@Test
+	public void shouldStopTestTask() throws Exception {
```
shouldStop or shouldShutdown?

suthagar23 replied:
renamed to shouldShutdownTestTask();

dkayiwa commented on omod-1.8/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs1_8/TaskActionController1_8Test.java:
```diff
@@ -0,0 +1,137 @@
+	}
+	
+	@Test
+	public void shouldDoNothingIfTestTaskAlreadyStarted() throws Exception {
```
What should do nothing?

suthagar23 replied:
That means, When I call "SheduleTask" action for the already scheduled task, It will not affect the status of that task.
Doing this test just for the confirmation.

dkayiwa commented on omod-1.8/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs1_8/TaskActionController1_8Test.java:
```diff
@@ -0,0 +1,137 @@
+	}
+	
+	@Test
+	public void shouldDoNothingIfTestTaskAlreadyStopped() throws Exception {
```
What should do nothing?

suthagar23 replied:
same as above!
renamed to shouldDoNothingIfTestTaskAlreadyShutdowned();

dkayiwa commented on omod-1.8/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs1_8/TaskActionController1_8Test.java:
```diff
@@ -0,0 +1,137 @@
+	}
+	
+	@Test
+	public void shouldUnloadTestTask() throws Exception {
```
What should unload test task?

suthagar23 replied:
I used this method to confirm the task deletion. So I named as shouldUnlodTestTask. 
Is it wrong?

dkayiwa replied:
If the test is about ensuring that the delete action should actually delete the task, then the name could be something like: shouldDeleteTask()

suthagar23 replied:
renamed :-)

dkayiwa commented on omod-1.8/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs1_8/TaskActionController1_8Test.java:
```diff
@@ -0,0 +1,137 @@
+	@Override
+	@Test(expected = Exception.class)
+	public void shouldGetAll() throws Exception {
+		super.shouldGetAll();
```
Does this mean that we do not support getting all tasks?

suthagar23 replied:
yah It will not support to that

dkayiwa replied:
Oh you are right, since these are just task actions. 😊

dkayiwa commented on omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_8/TaskDefinitionResource1_8.java:
```diff
@@ -0,0 +1,167 @@
+import org.openmrs.module.webservices.rest.web.resource.impl.DelegatingResourceDescription;
+import org.openmrs.module.webservices.rest.web.response.ResponseException;
+
+import java.util.*;
```
Take a look at this for the above: https://wiki.openmrs.org/display/docs/Java+Conventions#JavaConventions-Donotusewildcardimports

suthagar23 replied:
Refactored!

dkayiwa commented on omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_8/TaskDefinitionResource1_8.java:
```diff
@@ -0,0 +1,167 @@
+	@Override
+	public TaskDefinition save(TaskDefinition taskDefinition) {
+		taskServiceWrapper.saveTaskDefinition(taskDefinition);
+		return Context.getSchedulerService().getTaskByName(taskDefinition.getName());
```
Why do you fetch the task again?

suthagar23 replied:
No idea at that time, Just return taskDefinition is enough :-(

dkayiwa commented on omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_8/TaskDefinitionResource1_8.java:
```diff
@@ -0,0 +1,167 @@
+ */
+@Resource(name = RestConstants.VERSION_1 + "/taskdefinition", supportedClass = TaskDefinition.class, supportedOpenmrsVersions = {
+        "1.8.*", "1.9.*", "1.10.*", "1.11.*", "1.12.*", "2.0.*", "2.1.*" })
+public class TaskDefinitionResource1_8 extends MetadataDelegatingCrudResource<TaskDefinition> {
```
Do we have TaskDefinitionController1_8Test?

suthagar23 replied:
Updated :blush:  Now It contains TaskDefinitionController1_8Test also

dkayiwa replied:
How about TaskDefinitionResource1_8Test

suthagar23 replied:
It is also included in the PR. You can get it from here,
https://github.com/suthagar23/openmrs-module-webservices.rest/blob/0e7b6149d51960708f75419c210a54bdd2b93f54/omod-1.8/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_8/TaskDefinitionResource1_8Test.java

dkayiwa commented on omod-common/src/main/java/org/openmrs/module/webservices/helper/TaskAction.java:
```diff
@@ -53,5 +53,9 @@ public Action getAction() {
-	
+
+	public void setAllModules(boolean allTasks) {
+		this.allTasks = allTasks;
```
Can't we just drop the all tasks property? Of what use is it?

suthagar23 replied:
Here I have some SchedulerService functions for allTasks, So if the request contains `allTasks = true `then I will call that function to perform the actions for all tasks (like resheduleAllTasks). 
If the request doesn't contain `allTasks` property, then it will be false (initialization default value)

dkayiwa replied:
Do you mean is someone sends a list of tasks to be acted upon, then need to include the allTasks flag? If allTasks = false, does it mean that the action is just some of the passed in tasks but not others?

suthagar23 replied:
For Example, If the system has 5 tasks and he sent a list of 3 tasks (without allTasks property), then the resource will do that respected action for only that 3 tasks.

If he wants to perform that action for all 5 tasks, then he can just send action type with `allTasks=true` property. Then resource will not think about the list of tasks in the request. 
if will perform the required action for all tasks on the server.

dkayiwa replied:
Thanks for the clarification. 😊

suthagar23 commented:
@dkayiwa Completed :blush: Can you please take a look to merge!

dkayiwa commented on omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_8/TaskActionResource1_8.java:
```diff
@@ -0,0 +1,204 @@
+		}
+		if (!action.isAllTasks()) {
+			for (TaskDefinition taskDef : taskDefinitions) {
+				if (taskDef == null) {
```
Can a collection contain a null value?

suthagar23 replied:
I have removed this Condition, because the wrong TaskDefinition will be caught by the setConvertedProperties() and If there are any issues, then It will throw the error.

dkayiwa commented on omod-1.8/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs1_8/TaskActionController1_8Test.java:
```diff
@@ -0,0 +1,145 @@
+	}
+	
+	@Test
+	public void shouldScheduleTestTask() throws Exception {
```
No need to add "Test" to any of these test methods. For instance, shouldScheduleTask is enough instead of shouldScheduleTestTask

suthagar23 replied:
Refactored

dkayiwa commented on omod-1.8/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs1_8/TaskDefinitionController1_8Test.java:
```diff
@@ -0,0 +1,165 @@
+	}
+	
+	@Test
+	public void shouldGetModuleByUuid() throws Exception {
```
Should get what?

dkayiwa commented on omod-1.8/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs1_8/TaskDefinitionController1_8Test.java:
```diff
@@ -0,0 +1,165 @@
+	
+	@Test
+	@Override
+	public void shouldGetDefaultByUuid() throws Exception {
```
When you say shouldGet, add what exactly you are getting.

suthagar23 replied:
Here I have @Override the method from the super class. that will express the Default representation and I can't change the name here

dkayiwa commented on omod-common/src/main/java/org/openmrs/module/webservices/helper/TaskServiceWrapper.java:
```diff
@@ -0,0 +1,102 @@
+/**
+ * It will used to retrieve the required data from the Scheduler Service
+ */
+public class TaskServiceWrapper {
```
What value does this class add?

suthagar23 replied:
Wrapping the required task information between related resource and SchedulerService

dkayiwa replied:
This wrapping is simply one method call hence rendering it useless.

suthagar23 replied:
yah, but in the flow I want this class to create another Mock class for testing. 
Otherwise, I can't do the testing. 
I am using **TaskServiceWrapper** for active service works 
**MockTaskServiceWrapper** for mock testing works

dkayiwa replied:
You can always mock Context.getSchedulerService().getTask(), etc without writing new mock classes.

suthagar23 replied:
Here I followed the same way of the ModuleActionResource and ModuleWrapper and achieve the target :blush: 
What do you mean by the mocking Services?

dkayiwa replied:
Ok, let us go with what you have. 😊

suthagar23 commented:
@dkayiwa So Finally If you haven't any concerns or comments, Could we move to merge it :blush:

dkayiwa commented on omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_8/TaskActionResource1_8.java:
```diff
@@ -0,0 +1,196 @@
+	
+	@Override
+	public TaskAction getByUniqueId(String uniqueId) {
+		throw new UnsupportedOperationException("TaskAction cannot get by id");
```
Shouldn't we take this to be the internal id, if a number, or task name?

dkayiwa replied:
Oh sorry. Just ignore the above. 😊

suthagar23 replied:
Here It is the integer number.  Here I have overridden that method in TaskAction. 
Anyway, TaskAction is not supported to that function.

dkayiwa commented on omod-1.8/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs1_8/TaskActionController1_8Test.java:
```diff
@@ -0,0 +1,145 @@
+	
+	@Override
+	@Test(expected = Exception.class)
+	public void shouldGetDefaultByUuid() throws Exception {
```
should get what?

suthagar23 replied:
Overridden method. It is also like used to get Default Representation. But not supported to TaskAction.

dkayiwa commented on omod-common/src/main/java/org/openmrs/module/webservices/helper/TaskServiceWrapper.java:
```diff
@@ -0,0 +1,102 @@
+	 * @throws SchedulerException - It will throw in case of any SchedulerService exceptions
+	 */
+	public void deleteTask(TaskDefinition task) throws SchedulerException {
+		if (Context.getSchedulerService().getTaskByName(task.getName()).getStarted()) {
```
What is the use of this logic? Isn't it taken care of by the service layer? Why don't you just Context.getSchedulerService().deleteTask?

suthagar23 replied:
Nope, SchedulerService is allowed to delete only the shutdown tasks. Otherwise, It threw the error. (I think, Core doesn't care about the task status)
So I just check about the task status before delete that task

dkayiwa replied:
Even then, that is the business of the service layer. Your duty is to simply call Context.getSchedulerService().deleteTask

suthagar23 replied:
Okay :blush:

dkayiwa commented on omod-1.8/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs1_8/TaskDefinitionResource1_8Test.java:
```diff
@@ -0,0 +1,70 @@
+		assertPropEquals("started", getObject().getStarted());
+		assertPropEquals("taskClass", getObject().getTaskClass());
+		assertPropEquals("lastExecutionTime", getObject().getLastExecutionTime());
+		assertPropPresent("startTimePattern");
```
Shouldn't the above also be assertPropEquals()?

suthagar23 replied:
changed

dkayiwa replied:
How about the one below it? 😊

suthagar23 replied:
That too changed :blush:

dkayiwa commented on omod-1.8/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs1_8/TaskDefinitionController1_8Test.java:
```diff
@@ -0,0 +1,165 @@
+		MockHttpServletRequest req = request(RequestMethod.POST, getURI());
+		req.setContent(json.getBytes());
+		deserialize(handle(req));
+		Assert.assertEquals(mockTaskServiceWrapper.getTaskByName("MockTask").getName(), "MockTask");
```
How sure can i be that this task did not exist before saving?

suthagar23 replied:
I will add another test before to this to confirm that

dkayiwa commented on omod-1.8/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs1_8/TaskDefinitionController1_8Test.java:
```diff
@@ -0,0 +1,165 @@
+	}
+	
+	@Test
+	public void shouldGetAllRegistered() throws Exception {
```
What is the difference between shouldGetAll and shouldGetAllRegistered?

suthagar23 replied:
Defaultly GetAll method in the resource has been set to return the list of Scheduled tasks. (I can't change the name because it is an Overriden method)
So I am using another separate method to test about the Registered tasks.

dkayiwa replied:
Isn't it more natural for GetAll to return all tasks regardless of whether they are scheduled or not?

suthagar23 replied:
So I will change the way to retrieve all the registered for the getAll method

dkayiwa commented on omod-1.8/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs1_8/TaskActionController1_8Test.java:
```diff
@@ -0,0 +1,145 @@
+	}
+	
+	@Test
+	public void shouldDoNothingIfTaskAlreadyShutdown() throws Exception {
```
This should be: shutdown_shouldDoNothingIfTaskAlreadyShutdown

suthagar23 replied:
Changed

dkayiwa commented on omod-1.8/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs1_8/TaskActionController1_8Test.java:
```diff
@@ -0,0 +1,145 @@
+	}
+	
+	@Test
+	public void shouldDoNothingIfTaskAlreadyScheduled() throws Exception {
```
scheduleTask _shouldDoNothingIfTaskAlreadyScheduled

suthagar23 replied:
Changed

dkayiwa commented on omod-1.8/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs1_8/TaskActionController1_8Test.java:
```diff
@@ -0,0 +1,145 @@
+		mockTaskServiceWrapper.scheduledTasks.add(testTask);
+		deserialize(handle(newPostRequest(getURI(), "{\"action\": \"shutdowntask\", \"tasks\":[\"" + getTestTaskName()
+		        + "\"]}")));
+		assertThat(mockTaskServiceWrapper.scheduledTasks, not(hasItem(testTask)));
```
How how are you that the task wasn't already shutdown even before the shutdown command?

suthagar23 replied:
Here I am checking the Scheduled list before making shout down the request. The scheduled list will contain this task If it is not shut down. 
After the request, It will be removed from the Scheduled list (shutdown). Then I am checking for removal.

dkayiwa replied:
Oh which line number are you checking that the task is not already shutdown?

suthagar23 replied:
Ohh, Sorry really missed that :cry: 
I will add it quickly

suthagar23 commented:
Thank you so much @dkayiwa  for the review and merge :blush:

dkayiwa commented:
Thank you too for responding to all my back and forth review comments! 😊
